Image block: Lightbox animation improvements

artemiomorales opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
This PR will do the following in order to improve the UX around the lightbox animation:
1. On click, use the existing image if the full one hasn't loaded yet
2. Once the new one has loaded, replace it
3. Preload the full image on hover


## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Addresses #51555
Currently, there is a delay when the user clicks on an image to activate the lightbox. This will improve that experience.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
I've edited the code so that we now have two `img` elements inside of the lightbox. The lower resolution one is a copy of the image inside of the body content; there is also a larger resolution one whose `src` attribute is populated dynamically when when the lightbox is opened and the full-sized image finishes loading.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->
*Must be tested on a live site (not a local environment)
1. Make sure the Interactivity API experiments are enabled
2. Add an image block to a post and activate the lightbox under Advanced > Behaviors in the block settings
3. Publish the post and click on the image
4. The lightbox should respond immediately
5. If you'd like, you can inspect the source to ensure the high resolution image is the one being displayed in the lightbox

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
N/A

## Screenshots or screencast <!-- if applicable -->

### Before

https://github.com/WordPress/gutenberg/assets/5360536/b030b40a-8dbc-477f-a064-8a778c829dd5

### After

https://github.com/WordPress/gutenberg/assets/5360536/812ada30-a008-4bb7-a6d8-16ed9020558e

artemiomorales committed:
Add logic to use low-res image while high-res one is loading

We need to have two <img> elements in the DOM inside the lightbox
because otherwise the image flickers when we change the src.

I removed the src and srcset attributes from the responsive image
when the larger one is loaded to signal that the low-res one is no
longer in use.

artemiomorales commented:
I still need to do more testing to make sure I didn't inadvertently break the fade animation, as well as add the preloading of the image on hover.

While I'm at it, I'd also like to add a slight zoom on hover to improve the UX.

Lastly, I've spotted a couple of issues at different responsive sizes where the image gets distorted, so I'd like to address those, though perhaps that can be a separate PR.

artemiomorales committed:
Add logic to preload image on hover

artemiomorales commented:
> While I'm at it, I'd also like to add a slight zoom on hover to improve the UX.
>
> Lastly, I've spotted a couple of issues at different responsive sizes where the image gets distorted, so I'd like to address those, though perhaps that can be a separate PR.

I decided it makes more sense to handle these as part of a separate PR for easier review.

With that being said, I believe this PR is ready for feedback.

artemiomorales committed:
Update tests

cbravobernal committed:
PHP format

cbravobernal commented:
I did a quick PHP format. Let me know if you need any help setting auto-format in your environment.

artemiomorales commented on lib/block-supports/behaviors.php:
```diff
@@ -87,10 +87,18 @@ function gutenberg_render_behaviors_support_lightbox( $block_content, $block ) {
 	} else {
-		$img_src = $z->get_attribute( 'src' );
+		$img_src        = $z->get_attribute( 'src' );
+		$img_dimensions = getimagesize( $img_src );
```
@c4rl0sbr4v0 A note related to performance: I realize this line will fetch an entire image to determine dimensions when an image is inserted via link. We can likely find this information without needing to download the whole image, but that seems beyond the scope of this PR and an enhancement.

What do you think?

cbravobernal replied:
As you mention, could be out of the scope. But is still an important enhancement. As, for example, in an online newspaper, where a post can contain several images, we can cause unexpected performance problems or increase transfer costs to hosting providers.

cbravobernal replied:
There is a `wp_getimagesize()` function. Maybe we should use it as an alternative for consistency.
https://developer.wordpress.org/reference/functions/wp_getimagesize/

artemiomorales replied:
Done 👍

cbravobernal commented on lib/block-supports/behaviors.php:
```diff
@@ -99,24 +107,54 @@ function gutenberg_render_behaviors_support_lightbox( $block_content, $block ) {
+			}',
+			$lightbox_animation,
+			$img_src,
+			$img_srcset ? $img_srcset : '',
```
Had to change this, as null coalescing operator (??) is not present in PHP version 5.6 😢

artemiomorales replied:
Great, thanks!

artemiomorales committed:
Prevent responsive images from being loaded unnecessarily

The src attribute on the img elements was causing the srcset
attribute to be added as well before the Interactivity API
could remove them, causing images to be loaded before they
were needed and in the wrong size.

This commit removes the src attribute from the img elements
before they are output to the DOM, and also updates the tests.

artemiomorales commented:
@c4rl0sbr4v0 I just pushed up a change that should prevent responsive images from getting loaded unnecessarily 👍

cbravobernal commented:
Updated with trunk to see if we solve the e2e testing issues.

cbravobernal reviewed (approved):
LGTM and works as expected.

cbravobernal commented on lib/block-supports/behaviors.php:
```diff
@@ -99,24 +107,56 @@ function gutenberg_render_behaviors_support_lightbox( $block_content, $block ) {
+			}',
+			$lightbox_animation,
+			$img_src,
+			$img_srcset ? $img_srcset : '',
```
This line causes a warning if we use an image from URL:
![Screenshot 2023-06-23 at 12 05 37](https://github.com/WordPress/gutenberg/assets/37012961/508fe706-5cac-48d9-8097-f607d31c4514)

cbravobernal committed:
Prevent warning of undefined variable

cbravobernal committed:
Prevent warning of undefined variable - refactored

artemiomorales committed:
Replace getimagesize() with wp_getimagesize()

artemiomorales committed:
Resolve image flash when opening lightbox

Rather than allowing the browser to pick the lightbox's
responsive image, we now explicitly read the currentSrc
attribute from the reference image and also prevent users
from opening the lightbox until the reference image is
fully loaded.

Doing this, we can ensure the zoom animation plays smoothly
and there are no oddities in the UX.

artemiomorales commented:
@c4rl0sbr4v0 I just updated the logic to hopefully finally resolve the image flickering or appearing broken when opening the lightbox on production environments. Would you be able to give it a once-over to make sure it's still working as expected? (To simulate the latency, you could try testing with Chrome Dev Tools throttling enabled.)

If all looks good, I think we can merge. Let me know what you think!

cbravobernal commented on packages/block-library/src/image/interactivity.js:
```diff
@@ -22,6 +22,11 @@ store( {
 		core: {
 			image: {
 				showLightbox: ( { context, event } ) => {
+					// We can't initialize the lightbox until the reference
+					// image is loaded, otherwise the UX is broken.
+					if ( ! context.core.image.imageLoaded ) {
+						return;
+					}
```
I hope time to interactive does not being affected here too much.

artemiomorales replied:
It depends on the user's bandwidth. But also, I think it's reasonable to prevent a user from zooming in on an image that hasn't fully loaded yet? 🤔

Another note: We can likely make the time to interactivity instant with the fade animation; it's just the zoom animation that causes the trouble.

If it's worth exploring, I could make this initialization step specific to the zoom animation.

Another note: Is the concern here that users may want to go immediately to the zoomed image without needing to wait for the rest of the page to load? @mtias has [mentioned before](https://core.trac.wordpress.org/ticket/57913#comment:27) the idea of having a separate URL for the zoomed version of the image inside of the lightbox as an overall usability improvement for WordPress sites.

Perhaps we can continue to explore this in a subsequent PR where we start to identify how this would behave, (what should the URL be, how to implement the experience of initializing the page with the lightbox already open, crossover or not with the Media page, etc.)

cbravobernal replied:
> It depends on the user's bandwidth. But also, I think it's reasonable to prevent a user from zooming in on an image that hasn't fully loaded yet? 🤔

It is indeed. But, for example, IGDB on their Lightbox, shows a loading text if the image is not fully loaded. Still, is something that we can explore in future PRs. 

> Is the concern here that users may want to go immediately to the zoomed image without needing to wait for the rest of the page to load?

Yes, this could happen. Usually, clicking on an image and wait for the zoom to appear is not ideal.

We can wait and see how that ticket is evolving.

luisherranz commented on packages/block-library/src/image/interactivity.js:
```diff
@@ -142,6 +158,30 @@ store( {
+						} );
+					}
+				},
+				preloadLightboxImage: ( { context, ref } ) => {
+					ref.addEventListener( 'mouseover', () => {
```
Hey @artemiomorales, out of curiosity, why did you use `useEffect` instead of `data-wp-on--mouseover`?

artemiomorales replied:
Thanks for catching this! It's an oversight on my part. Will revise in this [upcoming PR](https://github.com/WordPress/gutenberg/pull/51823).

luisherranz replied:
Oh, no problem. Thanks, Artemio.

artemiomorales replied:
@luisherranz Update: I decided to break this out into a [separate PR](https://github.com/WordPress/gutenberg/pull/52067) for easier review and approval.
